rfc/egress-with-ucan.md

```diff
@@ -0,0 +1,237 @@
+    }
+  ],
+  "nnc": …,
+  "exp": …,
+  "prf": […]
+}
+```
+
+The UCAN 1.0 version of these delegations SHOULD specify an explicit `null` policy for the token, rather than include no policy: no policy would match both a `null`/missing token *and* any token, which almost certainly not intended.
+
+The Gateway MAY refuse to serve a request on the authority of an unchecked `token` field. This prevents a malicious actor from generating unexpected egress-incurring requests by including an invented token which will trigger the egress billing. The primary interface for creating these delegations SHOULD provide no means to create such a delegation; given that, the Gateway does not strictly need to check this, as the customer would have to go through some effort to make such a delegation.
+
+The delegation must be available to the Gateway at request-time, so it must have access to it in a store. The authorizing Client should therefore invoke `access/delegate` (UCAN 1.0 equivalent TBD) to store the delegation with Storacha.
+
+## The Gateway
+
+To retrieve a piece of content over HTTP, the HTTP client will connect to the existing [Storacha Gateway](https://w3s.link/) (also known as the web3.storage Gateway). The Gateway offers an HTTP interface which conforms to the [IPFS Subdomain Gateway Spec](https://specs.ipfs.tech/http-gateways/subdomain-gateway/). We extend that spec as follows:
+
+### Request Headers
+
+#### `Authorization` (request header)
+
+Optional. Provides a means of authorizing the content read. Follows the normal [HTTP `Authorization` spec](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization). The authentication scheme must be [`Bearer`](https://datatracker.ietf.org/doc/html/rfc6750). The token is the authorization token the Gateway should use.
+
+```text
+Authorization: Bearer abc123def456
+```
+
+### Request Query Parameters
+
+#### `auth-token` (request query parameter)
+
+Optional. Provides a means of authorizing the content read. The value is the authorization token the Gateway should use.
+
+```text
+https://bafybeidd2gyhagleh47qeg77xqndy2qy3yzn4vkxmk775bg2t5lpuy7pcu.ipfs.w3s.link/?auth-token=abc123def456
+```
+
+Requesting clients SHOULD NOT provide both an `Authorization` header and an `auth-token` query parameter. The Gateway MAY use either token if it does, leading to undefined behavior.
+
+### Serving a request
+
+To serve the request, the Gateway will perform the following steps:
+
+1. Look up the Location Commitment(s) for the given root CID from the Indexing Service.
+2. If none are found, respond with `404 Not Found`.
+3. Note if any Location Commitments were found with no Spaces. (If so, these are from before these changes, and mean we should fall back to the previous behavior later.)
+4. Get the set of unique Spaces from those Location Commitments. (There will usually be one Location Commitment, with one Space.)
+5. Repeat with each Space in any order, stopping if a successful response is produced:
```

Comment:
Is this going to potentially be slow? Is it a separate network request for each space we need to check? The performance is the only part of this spec I have concerns about, but I might be misunderstanding - has been a week or so since my head's been here so apologies if I'm missing something!